`isolatedDeclarations` false positive TS9013 on arrow functions with default parameters and `as const` objects

This issue was discovered on the https://github.com/react-querybuilder/react-querybuilder repo.

## Steps to reproduce

```ts
// repro.ts
const imported = ',';

export const fn1 = (x: string = imported): string => x;

interface Opts { val?: boolean }
export const fn2 = ({ val = false }: Opts = {}): boolean => val;

export const obj1 = { $and: [{ $expr: true }] } as const;
```

```jsonc
// tsconfig.json
{
  "compilerOptions": {
    "noEmit": true,
    "declaration": true,
    "isolatedDeclarations": true,
    "strict": true,
    "target": "esnext",
    "module": "esnext",
    "moduleResolution": "bundler"
  },
  "include": ["repro.ts"]
}
```

## Behavior with `typescript@6.0.2`

```
$ tsc -p tsconfig.json
// No errors
```

## Behavior with `tsgo` (7.0.0-dev.20260324.1)

```
$ tsgo -p tsconfig.json
repro.ts:3:21 - error TS9013: Expression type can't be inferred with --isolatedDeclarations.
repro.ts:6:21 - error TS9013: Expression type can't be inferred with --isolatedDeclarations.
repro.ts:8:29 - error TS9013: Expression type can't be inferred with --isolatedDeclarations.
repro.ts:8:39 - error TS9013: Expression type can't be inferred with --isolatedDeclarations.

Found 4 errors in the same file, starting at: repro.ts:3
```

## Additional context

These appear to be regressions from #2459 (isolatedDeclarations port, merged 2026-03-20). The same class of false positive was previously reported and fixed in tsc:

- [TypeScript#60192](https://github.com/microsoft/TypeScript/issues/60192) — false positive on arrow functions with typed optional/default parameters
- [TypeScript#60195](https://github.com/microsoft/TypeScript/pull/60195) — the fix

Two distinct patterns are affected:
1. **Arrow functions with default parameters** — `export const fn = (param: Type = expr): ReturnType => ...` is flagged even though the parameter already has a type annotation
2. **`as const` with nested object/array literals** — `export const obj = { key: [{ nested: true }] } as const` is flagged

--- a/internal/checker/pseudotypenodebuilder.go
+++ b/internal/checker/pseudotypenodebuilder.go
@@ -295,9 +295,12 @@ func (b *NodeBuilderImpl) pseudoTypeEquivalentToType(t *pseudochecker.PseudoType
 				}
 			}
 		}
+		// handles freshness mismatches (e.g., fresh true vs regular true in as const)
+		if b.ch.getRegularTypeOfLiteralType(typeFromPseudo) == b.ch.getRegularTypeOfLiteralType(type_) {
+			return true
+		}
 		if typeFromPseudo.flags&TypeFlagsUnion != 0 && type_.flags&TypeFlagsUnion != 0 {
-			// handles freshness and `undefined` variant mismatches among union members, plus union comparison in general, since the unions may not be `==`
-			// identical due to aliasing and the like
+			// handles union comparison in general, since unions may not be `==` identical due to aliasing
 			if b.ch.compareTypesIdentical(typeFromPseudo, type_) == TernaryTrue {
 				return true
 			}

--- a/testdata/baselines/reference/compiler/isolatedDeclarationsAsConstNestedObjects.js
+++ b/testdata/baselines/reference/compiler/isolatedDeclarationsAsConstNestedObjects.js
@@ -1,0 +1,16 @@
+//// [tests/cases/compiler/isolatedDeclarationsAsConstNestedObjects.ts] ////
+
+//// [isolatedDeclarationsAsConstNestedObjects.ts]
+export const obj1 = { $and: [{ $expr: true }] } as const;
+
+
+//// [isolatedDeclarationsAsConstNestedObjects.js]
+export const obj1 = { $and: [{ $expr: true }] };
+
+
+//// [isolatedDeclarationsAsConstNestedObjects.d.ts]
+export declare const obj1: {
+    readonly $and: readonly [{
+        readonly $expr: true;
+    }];
+};

--- a/testdata/baselines/reference/compiler/isolatedDeclarationsAsConstNestedObjects.symbols
+++ b/testdata/baselines/reference/compiler/isolatedDeclarationsAsConstNestedObjects.symbols
@@ -1,0 +1,9 @@
+//// [tests/cases/compiler/isolatedDeclarationsAsConstNestedObjects.ts] ////
+
+=== isolatedDeclarationsAsConstNestedObjects.ts ===
+export const obj1 = { $and: [{ $expr: true }] } as const;
+>obj1 : Symbol(obj1, Decl(isolatedDeclarationsAsConstNestedObjects.ts, 0, 12))
+>$and : Symbol($and, Decl(isolatedDeclarationsAsConstNestedObjects.ts, 0, 21))
+>$expr : Symbol($expr, Decl(isolatedDeclarationsAsConstNestedObjects.ts, 0, 30))
+>const : Symbol(const)
+

--- a/testdata/baselines/reference/compiler/isolatedDeclarationsAsConstNestedObjects.types
+++ b/testdata/baselines/reference/compiler/isolatedDeclarationsAsConstNestedObjects.types
@@ -1,0 +1,13 @@
+//// [tests/cases/compiler/isolatedDeclarationsAsConstNestedObjects.ts] ////
+
+=== isolatedDeclarationsAsConstNestedObjects.ts ===
+export const obj1 = { $and: [{ $expr: true }] } as const;
+>obj1 : { readonly $and: readonly [{ readonly $expr: true; }]; }
+>{ $and: [{ $expr: true }] } as const : { readonly $and: readonly [{ readonly $expr: true; }]; }
+>{ $and: [{ $expr: true }] } : { readonly $and: readonly [{ readonly $expr: true; }]; }
+>$and : readonly [{ readonly $expr: true; }]
+>[{ $expr: true }] : readonly [{ readonly $expr: true; }]
+>{ $expr: true } : { readonly $expr: true; }
+>$expr : true
+>true : true
+

--- a/testdata/baselines/reference/submodule/conformance/constAssertions.js
+++ b/testdata/baselines/reference/submodule/conformance/constAssertions.js
@@ -236,7 +236,7 @@ declare let vv1: "abc";
 declare let vc1: "abc";
 declare let a1: readonly [];
 declare let a2: readonly [1, 2, 3];
-declare let a3: readonly [10, "hello", true];
+declare let a3: readonly [10, 'hello', true];
 declare let a4: readonly [1, 2, 3];
 declare let a5: number[];
 declare let a6: readonly number[];

--- a/testdata/baselines/reference/submodule/conformance/constAssertions.js.diff
+++ b/testdata/baselines/reference/submodule/conformance/constAssertions.js.diff
@@ -22,7 +22,16 @@
  declare let c3: 10;
  declare let c4: -10;
  declare let c5: 10;
-@@= skipped -78, +78 lines =@@
+@@= skipped -13, +13 lines =@@
+ declare let vc1: "abc";
+ declare let a1: readonly [];
+ declare let a2: readonly [1, 2, 3];
+-declare let a3: readonly [10, "hello", true];
++declare let a3: readonly [10, 'hello', true];
+ declare let a4: readonly [1, 2, 3];
+ declare let a5: number[];
+ declare let a6: readonly number[];
+@@= skipped -65, +65 lines =@@
      };
  };
  declare let q1: 10;

--- a/testdata/tests/cases/compiler/isolatedDeclarationsAsConstNestedObjects.ts
+++ b/testdata/tests/cases/compiler/isolatedDeclarationsAsConstNestedObjects.ts
@@ -1,0 +1,5 @@
+// @isolatedDeclarations: true
+// @declaration: true
+// @strict: true
+
+export const obj1 = { $and: [{ $expr: true }] } as const;

